iOS/UZTickets/src/scan/UZTScanViewController.m

```diff
@@ -0,0 +1,180 @@
+    if (![session canAddInput:input]) return NO;
+    
+    return YES;
+}
+
+- (void)setupLayerSessionWithAvailableObjects:(NSArray*)availableObjectTypes
+{
+    [self.viewModel didStartInitializingScanView];
+    
+    self.dispatchQueue = dispatch_queue_create("UZTScanQueue", DISPATCH_QUEUE_SERIAL);
+    
+    dispatch_async(self.dispatchQueue, ^{
+        NSArray* devices = [AVCaptureDevice devicesWithMediaType:AVMediaTypeVideo];
+        NSAssert(devices.count > 0, @"There are no device that can capture video input installed on this device");
+        
+        AVCaptureSession* session = [AVCaptureSession new];
+        self.session = session;
+        
+        AVCaptureMetadataOutput* output = [AVCaptureMetadataOutput new];
+        [output setMetadataObjectsDelegate:self queue:self.dispatchQueue];
+        output.metadataObjectTypes = nil;
+        self.output = output;
+        
+        AVCaptureDevice* device;
+        
+        for (AVCaptureDevice* cameraDevice in devices)
+        {
+            if (cameraDevice.position == AVCaptureDevicePositionBack)
+            {
+                device = cameraDevice;
+                if ([device hasTorch] && [device isTorchModeSupported:AVCaptureTorchModeAuto])
+                {
+                    NSError *error = nil;
+                    if ([device lockForConfiguration:&error])
+                    {
+                        device.torchMode = AVCaptureTorchModeAuto;
+                        [device unlockForConfiguration];
+                    }
+                    NSAssert(error == nil, @"lock for configuration failed with error: %@", error);
+                }
+                break;
+            }
+        }
+        NSAssert(device != nil, @"There is no back camera available on this device");
+        
+        NSError* inputError;
+        AVCaptureDeviceInput* input = [AVCaptureDeviceInput deviceInputWithDevice:device error:&inputError];
+        NSAssert(inputError == nil, @"An error occurred while try get AVCaptureDeviceInput: %@", inputError);
+        
+        [self.session beginConfiguration];
+        NSAssert([session canAddInput:input], @"Can't add camera device input to session");
+        [session addInput:input];
+        
+        NSAssert([session canAddOutput:output], @"Can't add barcode recognition media output");
+        [session addOutput:output];
+        
+        NSSet* availableOutputObjectTypes = [NSSet setWithArray:output.availableMetadataObjectTypes];
+        NSSet* filterObjectTypes = [NSSet setWithArray:availableObjectTypes];
+        NSAssert([filterObjectTypes isSubsetOfSet:availableOutputObjectTypes], @"There are object types that can't be handled");
+        
+        output.metadataObjectTypes = availableObjectTypes;
+        [self.session commitConfiguration];
+        
+        [self.viewModel didInitializeScanView];
+    });
+}
+
+- (void)viewDidLoad
+{
+    [super viewDidLoad];
+    
+    self.viewModel = [UZTScanViewModel new];
```

Comment:
It is. This will be removed as soon as will be implemented viewmodel injection from storyboard mechanism.